### PR TITLE
Synchronise functions in Bytes and String

### PR DESCRIPTION
**Pull request by dra27**
While looking through history for #9446, I did a diff on the interfaces for `Bytes` and `String`. This PR proposes various missing functions in each:

- `String.empty` and `String.cat` mirror functions in `Bytes` for completeness only
- `String.of_bytes` and `String.to_bytes` are the obvious aliases for `Bytes.to_string` and `Bytes.of_string` respectively
- `Bytes.split_on_char` replicates #626's `String.split_on_char` (given that we have `Bytes.concat`, it makes sense also to have this one)
- `String.get_` decoding functions from #1864 are now included - these were mentioned as being for a future PR

The following seem to be entirely acceptable differences:
- `Bytes.unsafe_of_string` and `Bytes.unsafe_to_string` are special cases for `Bytes` only.
- `String.blit` is an hangover from mutable-string days, so `String.blit_bytes` doesn't make sense
- `Bytes.extend` is fundamental mutable, so there's no sense in a version in `String`

As ever, `Changes` is updated with each commit in order to make @gasche (hopefully) smile.

### CODE REVIEW

File: stdlib/bytes.ml
Comment (dra27) [inline stdlib/bytes.ml:356]: This is literally a copy of `String.split_on_char`, since it lexically draws in Bytes.sub and Bytes.unsafe_get instead.

File: stdlib/string.ml
Comment (dra27) [inline stdlib/string.ml]: These are also literally the same as in bytes.ml

File: stdlib/string.ml
Comment (dra27) [inline stdlib/string.ml:267]: The runtime already has the correctly-named primitives for the `string_` versions (the names of these four primitives constitute the only difference in the code between the `String` and `Bytes` versions)

File: testsuite/tests/lib-string/binary.ml
Comment (dra27) [inline testsuite/tests/lib-string/binary.ml:1]: This is just a quickly-adapted version of `lib-bytes/binary.ml`

**Review by dra27 — COMMENTED** [review]
A few notes - although the diff looks quite large, virtually all of the code is copied-and-pasted without alteration, so the review process should be straightforward.

**Comment by gasche** [comment]
[Grumpy stdlib bad cop.]

Personally I am not in favor of copy-pasting code around, unless there is a clear need. A few remarks:

- Some functions basically do not duplicate anything (`empty`, `cat`), no problem

- Sometimes it looks like it would be easy to define the String version in termes of the Bytes version with some unsafe coercion in the middle; the `get_[u]int<N>` functions for example. I don't know why you are duplicating the implementation. That said, if they are easy to get from Bytes, why is it important to have them in String?

  (For `split_on_char` it is mildly more irritating to derive it from the String version, as it requires a final `List.map`. No strong opinion on whether we should have it on both sides.)

- In general we don't have conversion functions in two modules, because in general it can create cyclic dependencies. In this general case I think the right approach is to wonder which module logically comes "first", and add conversion functions in the other one. (Seq comes first, datastructure modules bring their own converters to and from Seq). I think this explains why the conversions (safe or not) are only in Bytes. No very strong opinion on whether this particular pair should respect this style; but I remark that moving between safe and unsafe conversion functions (or the other way around) is a not-uncommon thing for this particular data boundary, so maybe staying just on one side makes sense.

**Comment by dra27** [comment]
The copying and pasting is because we're not taking advantage of any tools to reduce it - it would be sensible if String, Bytes, StringLabels and BytesLabels were all generated from the same sources. Given that the duplication is extremely small (especially for those `get_` functions) I'd prefer refactoring of hardly-ever-touched code not to block an API decision, if that's OK?

The unsafe functions are there for good reason, but I don't like the idea that we ever advocate the _routine_ use of functions containing the word "unsafe" - I'd go on the basis that if the function makes sense over `string` and makes sense over `bytes` then both modules should have it.

I don't have a strong opinion one way or the other for the conversions, although the mathematical/philosophical argument that `Bytes` is more general than `String` feels unusual in this case, given that `String` clearly cannot be implemented without something which definitely behaves like `Bytes` (i.e. although a mutable string is more general than an immutable string, in real life clearly the mutable string came first?).

However, I put each alteration into its own commit precisely so that they could be cherry-picked (or not)

**Comment by mshinwell** [comment]
I'm not a great fan of copying and pasting code either, but I'm even less of a fan of spending time debugging obscure problems that arise because the optimiser has been lied to, by virtue of using some kind of "unsafe" coercion (whose meaning will vary depending on the case).

Unless the volume of code really is vast, which it is unlikely to be for `Bytes` and `String`, I would have thought that a couple of simple comments on each side (saying in which file the other copy occurs) is preferable than having to reason about complications involving the optimiser.

**Comment by gasche** [comment]
Apologies for being unclear in my message above, but when I wrote

> Sometimes it looks like it would be easy to define the String version in termes of the Bytes version with some unsafe coercion in the middle.

My suggestion was that the implementation of the functions *in the stdlib* could be written with the unsafe coercions, instead of being duplicated. For example in string.ml (if that makes sense):

```ocaml
let get_uint16 s i = Bytes.get_uint16 (Bytes.unsafe_of_string s) i
```

> I'm even less of a fan of spending time debugging obscure problems that arise because the optimiser has been lied to.

I don't think "being lied to" applies here. Unlike Obj.magic, these "unsafe" functions convert between objects that actually have the same representation. They are only unsafe from a control-over-mutation discipline (they break the convention that we do not mutate strings). If your optimizer assumes that bytes and string are distinct, and that strings are immutable, I think the optimizer is in the wrong. If it makes no such assumption then those "unsafe" coercion functions should be handled as either blackboxes or the identity, in either case they are not lying.

(In fact it's a bit more subtle than that given that the equality or inequality of those two abstract types is not exposed in the language: both models are valid extensions of the agnostic implementation we have.)

**Comment by hhugo** [comment]
js_of_ocaml is moving toward using a different memory representation for string and bytes. Because of that, I would be in favor of not relying too much on the bytes module to implement the string one (to avoid conversions between the two)

**Comment by yallop** [comment]
> If your optimizer assumes that bytes and string are distinct, and that strings are immutable, I think the optimizer is in the wrong. 

I'm interested to hear more about what you have in mind here, because the compiler already makes use of string immutability for optimization.  For example, the following program prints `XBC`:

```ocaml
   let s = "ABC" and t = "ABC" in
       (Bytes.unsafe_of_string s).[0] <- 'X';
       print_endline t
```

**Comment by dra27** [comment]
@gasche - I see what you mean now, commit pushed doing that (given that that involved transforming the code, it's potentially more error prone than the previous `yG`, but I did check it!)

**Comment by alainfrisch** [comment]
Is it obvious for everyone that adding functions to manipulate strings as binary data (all the `get_int*`) is a good step?  I've no objection to it, but I tend to see the `string`/`bytes` distinction as something more than just immutable/mutable.  Strings are intended to hold textual data (usually assumed to be encoded in utf-8 these days), and I tend to use `bytes` even for immutable binary blobs (this is not more problematic to me than using `array` in APIs even when the functions are not supposed to modify the arrays and this cannot be expressed in their type).  So while I think it's a good goal to make sure `Bytes` implements at least all polymorphic functions from `Array` (`bytes` is "just" a `char array` with an optimized representation), I don't personally put a lot of importance on keeping `String` and `Bytes` in sync.

**Comment by mshinwell** [comment]
Lying to the optimiser includes claiming that a value is immutable when in fact it is not.  The fact that the underlying representations might be the same is coincidental.  It is beneficial to make various optimisations on immutable data (for example knowing the contents of a string, or a tuple, or a record, etc) and that simply cannot be done if the mutability claims are not accurate.

Regarding @alainfrisch 's point, it seems fine to me to provide these operations on both `String` and `Bytes`, although I agree that in general these modules would not necessarily be expected to have the same interface.  I do not however agree on the point that things of type `string` would be expected to contain textual data.  It may be beneficial in some circumstances to obtain extra immutability guarantees   on data that is not human-readable.

**Comment by dra27** [comment]
@alainfrisch - no, not necessarily! I was willing to file it as being `Bytes`-specific, but it was only that it was noted several times in the original PR that the `String` version would be a follow-up that I added it in its own commit here.

**Comment by gadmm** [comment]
The documentation of `Bytes.unsafe_to_string` and `Bytes.unsafe_of_string` is clear. The user is responsible for respecting immutability assumptions through an ownership discipline.

What @gasche proposed looks correct to me for a slightly convoluted ownership argument:
```ocaml
let get_uint16 s i = Bytes.get_uint16 (Bytes.unsafe_of_string s) i
```
If you know that _`f` does not mutate its `Bytes.t` argument_, then you can always write:
```ocaml
let f_string s = f (Bytes.unsafe_of_string s)
```
f may receive an argument that has immutable aliases, which it does not interfere with by definition. What would be broken for an optimiser would be to introduce mutations in code that does not mutate! (Strictly so, from an ownership argument, because the optimiser does not know that it owns the Bytes.t value, and so the mutation is allowed to be observed in a different thread, even if the optimiser undoes the mutations later.)

The converse is more delicate, and that one would require the assumption that strings are mutable:
```ocaml
let f_bytes b = f (Bytes.unsafe_to_string b)
```
As the documentation points-out, this is only valid if the caller has single ownership of s (so that it can assume that the string is immutable):
>    The caller may not mutate [the argument] while the string is borrowed (it has temporarily given up ownership). This affects concurrent programs,   but also higher-order functions: if {!String.length} returned   a closure to be called later, [the argument] should not be mutated until this   closure is fully applied and returns ownership.

In other words, an API that defines a function in this way _must_ indicate that it can only be called if the caller has single ownership of the argument. That makes it rather less useful than a function that would duplicate the code, but the need for code duplication is easy to understand if one looks at the generated code: one version can be more optimised based on the immutability assumption.

I would say that this is one case where code duplication is warranted (and where templates would do better because they automate the duplication of code). One must be careful about maintaining the coherence between implementations, but `bytes.ml` already says:
> (* WARNING: Some functions in this file are duplicated in string.ml for
   efficiency reasons. When you modify the one in this file you need to
   modify its duplicate in string.ml.
   These functions have a "duplicated" comment above their definition. *)

So, if no benefit from optimisation is expected for `f_string` knowing that the argument is immutable (edit: taking into account the fact that these benefits could be brought back by the optimiser via inlining), and it is part of the contract that `f` does not leak or mutate its argument, then it is much better to define the string version from the bytes version as proposed by @gasche. (Functions that use `f_string` are still correct in optimising knowing that the argument is immutable, because immutability is preserved.)

**Comment by Drup** [comment]
Why not write `CamlInternalBytes` functorized in term of the string primitives (create, length, get, set), implement everything there, and ensures the inliner does its job (`[@inline always]` is respected by the non-flambda inliner nowadays, no?). It'll even be useful later, when/if we get @lpw25 's monomorphic arrays.

**Comment by gadmm** [comment]
@hhugo 

The zero-cost convertibility between bytes and string is part of the design and documentation of the Bytes and String modules, and the documentation indicates several safe ways users can leverage this convertibility (for instance to create a string using mutations before publishing it as an immutable value). To me it seems wrong to treat it only as an implementation detail. If `js_of_ocaml` gave up the zero cost, then I think users would be correct if they complained that `js_of_ocaml` introduces buggy behaviour in correct OCaml programs (depending on what promises are made to users).

(Off-topic for the current PR; I can repost this elsewhere if you tell me where to continue the discussion on the js_of_ocaml repository:) The problem of immutability vs. expressiveness in programming languages is _far_ from being an obvious topic. When OCaml decided to introduce immutable strings, it decided for an informal ownership discipline where bytes and string are two sides of the same coin, because they had to retain expressiveness (to a reasonable degree). In retrospect I find that this was very ambitious, and trying to sensitize users to the idea of ownership is commendable. Javascript's design for strings is different, and less expressive. It is more correct to think of OCaml's “safe” strings not to be immutable, but to be mutable provided you know you own them, closer to Rust. This is _not_ a small difference. I wonder if something could be done to avoid fragmenting OCaml over string semantics. (edit: in the follow-up conversation I change my mind about the notion of owning a string, because it is so nearly impossible to concretise that it is better to assume that all strings are shared, but I still think the zero-cost conversion is important, and still worry about the impact of fragmenting OCaml over string semantics.)

**Comment by gadmm** [comment]
@yallop This behaviour would be against the documentation of `Bytes.unsafe_of_string`, so I want to open an issue. But I could not reproduce it with `ocamlopt -O3`. Can you please share your setting and command line options? (This is not about immutability, but aliasing.)

**Comment by yallop** [comment]
> The zero-cost convertibility between bytes and string is part of the design and documentation of the Bytes and String modules, and the documentation indicates several safe ways users can leverage this convertibility 

The documentation seems inconsistent on this point.

First, the `Bytes` documentation [indicates that a `bytes` value is a flat array](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L19-L20):
> Each byte can be indexed in constant time for reading or writing.

Second, the documentation [explicitly permits non-flat representations for strings](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L436-L438):
> For example, mutating [unsafe_of_string ("foo" ^ s)] could mutate the shared string ["foo"] -- assuming a rope-like representation of strings.

Finally, the documentation [indicates that `bytes` and `string` have the same representation](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L325), since it says of the `unsafe_*` functions:
> They do not copy the internal data;

**Comment by yallop** [comment]
@gadmm: This behaviour is the default in OCaml 4.10.0:
```
$ ocaml -version
The OCaml toplevel, version 4.10.0
$ cat xbc.ml 
let s = "ABC" and t = "ABC" in
    (Bytes.unsafe_of_string s).[0] <- 'X';
    print_endline t
$ ocamlopt xbc.ml 
File "xbc.ml", line 2, characters 4-41:
2 |     (Bytes.unsafe_of_string s).[0] <- 'X';
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Alert deprecated: Stdlib.String.set
Use Bytes.set instead.
$ ./a.out 
XBC
```

But I don't think it is against the documentation, which [says](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L421-L422):
> For example, string literals are implicitly shared by the compiler, so you never uniquely own them.

See also https://github.com/ocaml/ocaml/issues/6991.

**Comment by lpw25** [comment]
@gadmm `Bytes.unsafe_to_string` and `Bytes.unsafe_of_string`  were only intended to be temporary measures whilst the switch to an immutable string world was in progress. We should not enshrine their behaviour in the compiler and insist that our strings are all secretly mutable forever -- that would defeat some of the point of making them immutable in the first place -- we should just remove the explicitly marked `unsafe` functions from the standard library.

**Comment by gasche** [comment]
I don't agree with @lpw25; today in many situations creating strings is completely impractical without going through bytes for initialization -- and I fail to see a reason to remove the tool we use to avoid the performance impact of a copy. If we wanted to remove these unsafe functions, we would need to seriously rethink the String API. We could do try to do this (other people have proposed better APIs, if I remember correctly @dbuenzli has one library just for the String module), but this is more work than copying functions from one module to another just because.

**Comment by gasche** [comment]
People have very different opinions on how Bytes/String should go. Some people favor seeing string as a opaque user-level type for "text" data, other people insist that immutable byte sequences should be kept as first-class use-cases as well (when they could use `bytes` and be careful, just like we did before for strings). People have different opinions on whether it will be practical someday to use different representations in the OCaml compiler. I don't think this particular PR is the best place to have a discussion about this.

P.S.: we could open a separate issue to discuss it (or preferably: some small corner of the problem) if people insist, or encourage people to write RFCs if they have a "Big Plan" they want to share with the world.

**Comment by gadmm** [comment]
~I determined that the discussion about string semantics was on-topic for this PR, so I allow myself to continue the discussion here.~ (edit: I just read @gasche's message. Let's continue the discussion elsewhere if you want.)

> The documentation seems inconsistent on this point.
> 
> Second, the documentation [explicitly permits non-flat representations for strings](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L436-L438):
> 
> Finally, the documentation [indicates that `bytes` and `string` have the same representation](https://github.com/ocaml/ocaml/blob/9f95355acb42b98347253d0957b4ed808affe692/stdlib/bytes.mli#L325), since it says of the `unsafe_*` functions:
> 
> > They do not copy the internal data;

Not exactly: it does not forbid for instance having a dual representation for strings (bytes⊕js_string), and ~raising an exception when trying to cast a js_string on the grounds that ownership is violated~ (edit: one should be able to pass any string to a bytes function that does not mutate).

I also think the documentation is consistent if we understand it to say that `^` does not return ownership of the result (and so unsafe_of_string would be allowed to fail when called on `"foo"^s`, or flatten the string as a fail-safe to be realistically backwards-compatible).

About string literals, you are right, I missed this sharing clause. The program is correct.

It is unfortunate that the less ways one has to receive ownership of a string, the less the conversion is useful. Any evolution should be evaluated in light of concrete usage of the feature (opam packages offer a good number of examples).

@lpw25 My understanding is that “safe strings” were introduced as a result of the [LaFoSec security report](https://www.ssi.gouv.fr/agence/publication/lafosec-securite-et-langages-fonctionnels/) and the current specification reaches the goal, but I do not see the problems with immutability assumptions being relied on more generally, this is what the ownership restrictions afford you. The issue was rather about aliasing, but this has now been clarified by Jeremy. I disagree that one should break once more hundreds of programs, the current situation specification seems fine, and this particular brand of ownership might be strengthened in the future by the kind of type system @Drup is working on, if one wanted to further evolve the situation.

**Comment by dra27** [comment]
Rebased, just to squash in the non-code-duplicating versions of the get_* functions (irrespective of the other fine discussion, that seems better _now_ and then they could be changed later with further changes to the compiler's treatment of `bytes`/`string`).

@mshinwell's point about cross-referencing was already addressed (although it's now moot). I was not advocating with this PR that `Bytes` and `String` should _always_ be in sync, it was more that surveying the existing differences showed that in at least two cases the `Bytes` version had not been considered at all or the `String` versions were explicitly left as a TODO item. So it seemed worth a one-time sync and then continue from there. My final(ish) feelings:

- Future additions to `String` and `Bytes` should _consider_ whether the function should be added to both, and add both if there's no clear reason for the other not to exist (on which I differ from @gasche)
- `string` and `bytes` are a specialish case, unlike data structures, and given that you can sort of argue chicken-and-egg as to which one comes first, I'd prefer to have the conversion functions in both modules (but not intending any argument over two aliases!)

**Comment by yallop** [comment]
@gadmm:
> I disagree that one should break once more hundreds of programs

Given the ownership discipline you described above, very little would break if we simply aliased the unsafe versions to the safe versions (morally speaking; I realise that the dependency is currently the other way around) :
```ocaml
let unsafe_of_string = of_string
let unsafe_to_string = to_string
```

**Comment by gasche** [comment]
On the other side, I have mellowed out on the whole thing and I think the additions now are reasonable. I dislike `split_on_char`, but it's not objectively worse to have it in Bytes as well, and if it's the only thing that adds real code duplication in the PR I can forget about it to make @dra27 happy.
(P.S.: on a slightly more formal note, yes, I agree that consistency is also useful, and this PR might be a case of improving consistency with few downsides.)

( @gadmm I also suspect that this point might be best discussed elsewhere, but I am flattered by your praise for the documentation of the `unsafe_*` conversion functions, and it was very useful to realize that we have a problem coming up with different opinions of what is the "truth" about those we should teach to our optimizers. )

**Comment by gasche** [comment]
I created a meta-issue #9451 if people want to discuss the Bytes/String semantics.

**Comment by dra27** [comment]
I've updated this PR to reflect not being in 4.11. Is there anything approaching a consensus on parts/all of this?

**Comment by gasche** [comment]
Personally I like the `get_<foo>` functions and `empty`, I am why-not? about `cat`, unconvinced by `String.{of,to}_bytes`, and slightly irritated by `split_on_char`. If no one else is willing to make decisions, I can get out of my usual stdlib neutral stance and approve the subset with `get_<foo>`, `empty` and `cat`.

**Review by damiendoligez — APPROVED** [review]
After re-reading the whole discussion, I'm (like @gasche) unconvinced by `String.{to,of}_bytes` and slightly annoyed that we are introducing more calls to `bos`. But overall I still think this PR is good as a whole.

**Comment by hhugo** [comment]
@damiendoligez, I've a GPR to remove all usage of Bytes.unsafe_of_string (aka `bos`) in the stdlib at the cost of more code duplication https://github.com/ocaml/ocaml/pull/9789

**Comment by dra27** [comment]
I'm unclear on the consensus for this - can I rebase and merge?

**Comment by gasche** [comment]
My opinion remains as in my last comment https://github.com/ocaml/ocaml/pull/9448#issuecomment-623505359 . There are things in the PR which I suspect would be better left out, but I won't oppose merging in the current state. Given that Damien approved, I don't see an obstacle to merging.

File: stdlib/bytes.mli
Comment (damiendoligez) [inline stdlib/bytes.mli]:
```suggestion
    subarrays of [s] that are delimited by the [sep] character.
```
Since `bytes` are "byte arrays" I think we should avoid the "substring" terminology.

**Review by damiendoligez — APPROVED** [review]
My approval still stands, and this has been open for months without raising any objection. I say you can rebase and merge.

File: stdlib/bytes.mli
Comment (dra27) [reply stdlib/bytes.mli]: It felt strange to me writing subarrays - we describe them as "Byte sequence operations", so I went with subsequences if that's OK?

**Comment by dra27** [comment]
Rebased and updated for 4.13. In order to check what had changed, I added a `#sig_use` directive to the toplevel in https://github.com/dra27/ocaml/commit/aded686e990e559ff833187075fce1bc3f451851 and then a test to see whether the `String` and `Bytes` are equivalent in https://github.com/dra27/ocaml/commit/4f2a4668e6b3ad55186cec6842a410459c91adc8 (these can be pushed in a separate PR if wanted). It revealed that `Bytes.ends_with` and `Bytes.starts_with` don't exist:

```
> File "./string_bytes_equiv.ml", line 83, characters 41-47:
> 83 | module M(String: BYTES) = Bytes_includes(String)
>                                               ^^^^^^
> Error: Signature mismatch:
>        ...
>        The value `ends_with' is required but not provided
>        File "/home/dra/ocaml/stdlib/string.mli", lines 163-164, characters 0-79:
>          Expected declaration
>        The value `starts_with' is required but not provided
>        File "/home/dra/ocaml/stdlib/string.mli", lines 156-157, characters 0-79:
>          Expected declaration
```
so I added them here. Interestingly, the question was asked in the PR whether the `Bytes` functions should be added, but it never appeared to be answered... perhaps the test would therefore be useful, on the basis that it forces either the equivalent functions to be added, or the test to be updated to record that they're intentionally missing.

**Comment by johnwhitington** [comment]
If you have a moment to merge this soon, it would be appreciated.

I have just prototyped a script to generate bytes.ml and string.ml from a common source stringbytes.ml, and so dispense with all the duplication, but I'd like to wait until this is merged so I have a stable base.

**Comment by xavierleroy** [comment]
For the record: I'm not a big fan of what was done here, see https://github.com/ocaml/ocaml/pull/10296#issuecomment-800069417 .  Bytes is about byte arrays and String about character strings, and those have less and less to do with each other.

**Comment by dbuenzli** [comment]
> Bytes is about byte arrays and String about character strings, and those have less and less to do with each other.

Given the shift towards Unicode, `String.t` as exists is hardly about character strings. It's at most about byte-level encodings of characters strings. But byte-level encodings are not really what you want to work with if you are really serious about text munging. 

As such I think we should rather see `String.t` as immutable bytes arrays.

In practice the schism between `Bytes.t` and `String.t` and their discrepancy in functionality is a tad annoying whenever you treat your data right after input and don't want to incur the cost of copying your `Byte.t` to a `String.t` to use some of `String`'s functions. What that means is that you get into `Bytes.unsafe_to_string` territory to avoid this which is brittle in general.

I think that if we want to minimize the use of unsafe functions it's likely better to synchronize the immutable functionality of mutable and immutable byte arrays.

**Comment by dbuenzli** [comment]
> I think that if we want to minimize the use of unsafe functions it's likely better to synchronize the immutable functionality of mutable and immutable byte arrays.

However looking at the signatures of the functions of this PR more closely and trying to imagine alternatives, I'm not sure that's the right strategy either :-)

**Pull request merged**
